### PR TITLE
Add new key word for new apps

### PR DESCRIPTION
**Pull request by samaradel**
### Description

- Add label with ```New``` for new applications.
- Just update ```hasUpdate``` prop in the applications list to true and ```releaseDate``` to see changes.
- Add badge for new updates in sidebar.

### Changes

image
image

### Test cases

image
- return ```true``` if the ```releaseDate``` is within 30 days.
- return ```false``` if the ```releaseDate``` was from more than 30 days.
- return ```true``` if ```releaseDate``` is today.

### Related Issues

- https://github.com/threefoldtech/tfgrid-sdk-ts/issues/3179

### Checklist

- [ ] Tests included
- [x] Build pass
- [ ] Documentation
- [x] Code format and docstrings
- [x] Screenshots/Video attached (needed for UI changes)

### CODE REVIEW

File: packages/playground/src/components/applications/ApplicationCards.vue [hunk 1]
```diff
@@ -43,21 +58,93 @@ export default defineComponent({
...
     const baseURL = import.meta.env.BASE_URL;
+    const filteredCards = ref(props.cards);
+
+    function isCreatedToday(timestamp: number) {
+      const today = new Date();
+      const startOfDay = new Date(today.getFullYear(), today.getMonth(), today.getDate());
+      const endOfDay = new Date(today.getFullYear(), today.getMonth(), today.getDate(), 23, 59, 59, 999);
+
+      return timestamp >= startOfDay.getTime() && timestamp <= endOfDay.getTime();
+    }
```
Comment (Mahmoud-Emad) [inline packages/playground/src/components/applications/ApplicationCards.vue]: Can't you check if the `createdAt.day === today.day`?

File: packages/playground/src/views/solutions_view.vue [hunk 2]
```diff
@@ -30,33 +30,43 @@ export default {
           "Nostr is a simple, open protocol that enables global, decentralized, and censorship-resistant social media.",
         icon: "nostr.png",
         route: DashboardRoutes.Applications.Nostr,
+        isNew: false,
+        createdAt: 0, // Update creation time in milliseconds ex: 1723468996310 is 'Mon Aug 12 2024 16:23:16'
```
Comment (Mahmoud-Emad) [inline packages/playground/src/views/solutions_view.vue]: If you choose to save the createdAt field inside the card object, I guess you have to hard-code the date in place
```suggestion
        isNew: false,
        createdAt: 1723468996310,
```

**Review by Mahmoud-Emad — CHANGES_REQUESTED** [review]
Good job ya Samar

**Review by zaelgohary — COMMENTED** [review]
I tried to run the Dashboard multiple times but it lags on this page. I'm not sure why.

image


It happens only on this branch. I tried staging & different branches and they worked fine.

File: packages/playground/src/App.vue [hunk 3]
```diff
@@ -248,7 +250,20 @@ function navigateToHome() {
...
+      const next30DaysInMs = 30 * 24 * 60 * 60 * 1000;
+      const timestamp = 1724769030739; // Update with the time of creation
+      const next30DaysFromNow = timestamp + next30DaysInMs;
+
+      if (item.hasUpdate == true) {
+        item.hasUpdate = Date.now() < next30DaysFromNow;
+      }
+    });
+  });
+});
```
Comment (zaelgohary) [inline packages/playground/src/App.vue]: I think the issue I'm facing happens because of the change in onMounted. It worked once I've reverted the change.

File: packages/playground/src/views/solutions_view.vue [hunk 4]
```diff
@@ -30,40 +30,52 @@ export default {
           "Nostr is a simple, open protocol that enables global, decentralized, and censorship-resistant social media.",
         icon: "nostr.png",
         route: DashboardRoutes.Applications.Nostr,
+        isNew: false,
+        createdAt: 1725283777169, // Update creation time in milliseconds ex: 1723468996310 is 'Mon Aug 12 2024 16:23:16'
```
Comment (zaelgohary) [inline packages/playground/src/views/solutions_view.vue]: Instead of hard coding the timestamps, what if we created a function to create timestamps and call it here?

File: packages/playground/src/utils/types.ts [hunk 5]
```diff
@@ -21,4 +21,6 @@ export interface ApplicationCard {
   icon: string;
   route: string;
   tags?: string[];
+  isNew: boolean;
+  createdAt: number;
```
Comment (zaelgohary) [inline packages/playground/src/utils/types.ts]: Please make this optional. This way, you won't need to add it to the old solutions.

Comment (zaelgohary) [reply packages/playground/src/utils/types.ts]: The createdAt I mean.

File: packages/playground/src/App.vue [hunk 6]
```diff
@@ -253,6 +255,19 @@ onMounted(async () => {
   openProfile.value = true;
 });
 
+onMounted(() => {
+  routes.map(route => {
+    route.items.map(item => {
+      const next30DaysInMs = 30 * 24 * 60 * 60 * 1000;
+      const timestamp = 1724769030739; // Update with the time of creation
```
Comment (zaelgohary) [inline packages/playground/src/App.vue]: This shouldn't be hardcoded.

File: packages/playground/src/App.vue [hunk 7]
```diff
@@ -253,6 +255,19 @@ onMounted(async () => {
...
+      const next30DaysFromNow = timestamp + next30DaysInMs;
+      if (item.hasUpdate == true) {
+        item.hasUpdate = Date.now() < next30DaysFromNow;
+      }
+    });
+  });
+});
```
Comment (zaelgohary) [inline packages/playground/src/App.vue]: Also, can you provide docstrings for the logic in the onMounted here?

File: packages/playground/src/utils/types.ts (hunk 5 above)
Comment (samaradel) [reply packages/playground/src/utils/types.ts]: the old solutions could have releases too

File: packages/playground/src/views/solutions_view.vue (hunk 4 above)
Comment (samaradel) [reply packages/playground/src/views/solutions_view.vue]: I think it can't be dynamic as it will be called each time the component is mounted so it will get the date of the current day each time

**Comment by MohamedElmdary** [comment]
@samaradel can't we have a simple way and a stronger typescript for dates for example
we can pass `{  year: 2024, month: "Sep", day: 4 }` instead of 1725454606447 which makes no sense just looking at it without understanding the deep details about project?

**Comment by MohamedElmdary** [comment]
trying to understaind what is the usage of `hasUpdate` if the project has a date and that date within 30 days that enough to show `new` chip why do we need `hasUpdate` ?

**Comment by samaradel** [comment]
@MohamedElmdary good point 👉

File: packages/playground/src/App.vue [hunk 8]
```diff
@@ -253,6 +255,27 @@ onMounted(async () => {
...
+  const next30DaysInMs = 30 * 24 * 60 * 60 * 1000;
+  routes.map(route => {
+    route.items.map(item => {
+      const next30DaysFromNow = item.timestamp ? item.timestamp + next30DaysInMs : 0;
+      if (item.hasUpdate == true) {
+        item.hasUpdate = Date.now() < next30DaysFromNow;
+      }
+    });
+  });
+});
```
Comment (MohamedElmdary) [inline packages/playground/src/App.vue]: this logic should be handled in applicationCard itself

File: packages/playground/src/components/applications/ApplicationCards.vue [hunk 9]
```diff
@@ -1,10 +1,23 @@
...
-    <v-col v-for="card in cards" :key="card.title" cols="12" sm="12" md="12" lg="6" :xl="cards.length > 3 ? 4 : 6">
+    <v-col
+      v-for="card in filteredCards"
+      :key="card.title"
+      cols="12"
+      sm="12"
+      md="12"
+      lg="6"
+      :xl="filteredCards.length > 3 ? 4 : 6"
+    >
```
Comment (MohamedElmdary) [inline packages/playground/src/components/applications/ApplicationCards.vue]: please avoid style updates in code spots that has no code changes as it's very confusing

File: packages/playground/src/utils/types.ts [hunk 10]
```diff
@@ -21,4 +21,6 @@ export interface ApplicationCard {
   icon: string;
   route: string;
   tags?: string[];
+  hasUpdate: boolean;
```
Comment (MohamedElmdary) [inline packages/playground/src/utils/types.ts]: can't get why we need `hasUpdate` also why it's required ... if it's not a prop in object so it's false

File: packages/playground/src/App.vue (hunk 8 above)
Comment (samaradel) [reply packages/playground/src/App.vue]: This part of adding badges in the sidebar, how could it be handled from the application part?

Comment (MohamedElmdary) [reply packages/playground/src/App.vue]: ah my bad it's almost the same as application can we make a util which handling this logic if it's duplicated?

File: packages/playground/src/components/applications/ApplicationCards.vue (hunk 9 above)
Comment (samaradel) [reply packages/playground/src/components/applications/ApplicationCards.vue]: it's handling the card sizes across all screens

**Comment by amiraabouhadid** [comment]
Please fix build error

File: packages/playground/src/views/vms_view.vue [hunk 11]
```diff
@@ -26,6 +26,7 @@ export default {
           "Deploy a full virtual machine on Threefold Grid. Full VM allows you to have a complete image with a custom kernel optimized for your own usecase.",
         icon: "vm.png",
         route: DashboardRoutes.VirtualMachines.FullVirtualMachine,
+        // releaseDate: { year: 2024, month: 9, day: 4 },
```
Comment (Mahmoud-Emad) [inline packages/playground/src/views/vms_view.vue]: Is that happen by mistake? the commented code

Comment (samaradel) [reply packages/playground/src/views/vms_view.vue]: No, it's the format to remind who will edit how it should be like

Comment (0oM4R) [reply packages/playground/src/views/vms_view.vue]: I think should be added in docs or, in doc string

Comment (samaradel) [reply packages/playground/src/views/vms_view.vue]: I think @Mahmoud-Emad is right, I should remove ```releaseDate``` comment it's interface job so there is no use for the doc string too

Comment (0oM4R) [reply packages/playground/src/views/vms_view.vue]: > I think @Mahmoud-Emad is right, I should remove `releaseDate` comment it's interface job so there is no use for the doc string

Agree, so is it mentioned on  the interface docstring?  if not please add it

Comment (samaradel) [reply packages/playground/src/views/vms_view.vue]: done

File: packages/playground/src/utils/date.ts [hunk 12]
```diff
@@ -1,5 +1,15 @@
 import moment from "moment";
 
+import type { Date } from "./types";
 export default function toHumanDate(timeInSeconds: number): string {
   return moment(timeInSeconds * 1000).format("DD MMM YYYY, hh:mm A");
 }
+
+export function isReleasedOverMon(date: Date): boolean {
+  const releaseDate = new Date(date.year, date.month - 1, date.day);
```
Comment (amiraabouhadid) [inline packages/playground/src/utils/date.ts]: why do you re-assign the release date?
what if item.releaseDate is undefined as it is optional? Please wrap in try and catch

Comment (samaradel) [reply packages/playground/src/utils/date.ts]: I'm not reassigning it it's just a variable name. and it's inside if condition in the template

File: packages/playground/src/utils/types.ts [hunk 13]
```diff
@@ -15,10 +15,37 @@ export interface IPublicConfig {
...
 
+/**
+ * Represents a date object with day, month, and year properties.
+ *
+ * @interface Date
+ * @property {number} day - The day of the month (1-31).
+ * @property {number} month - The month of the year (1-12).
+ * @property {number} year - The year.
+ */
+export interface Date {
```
Comment (xmonader) [inline packages/playground/src/utils/types.ts]: Why is this needed ? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date

File: packages/playground/src/utils/date.ts [hunk 14]
```diff
@@ -1,5 +1,15 @@
...
+  const currentDate = new Date();
+  const next30Days = 24 * 60 * 60 * 1000;
```
Comment (xmonader) [inline packages/playground/src/utils/date.ts]: change the name to `ThirtyDaysInMilliseconds`

File: packages/playground/src/utils/date.ts (hunk 12 above)
Comment (xmonader) [reply packages/playground/src/utils/date.ts]: I don't understand this logic. Specially subtracting 1 from the month of the passed date?


The function should be `differenceIsOverAMonth(Date1, Date2)` and when used you pass the current date and the release date

File: packages/playground/src/utils/date.ts [hunk 15]
```diff
@@ -3,3 +3,27 @@ import moment from "moment";
...
+ * This function calculates the difference in days between the provided date and the current date.
+ * It then checks if this difference is less than or equal to 30 days.
+ *
+ * @example
+ * const date = new Date('2022-01-01');
+ * console.log(thirtyDaysInMilliseconds(date)); // Output: true or false depending on the current date
+ */
+
+export function thirtyDaysInMilliseconds(date: Date): boolean {
+  const releaseDate = new Date(date.getFullYear(), date.getMonth(), date.getDate());
```
Comment (xmonader) [inline packages/playground/src/utils/date.ts]: What is the point of this?

const releaseDate = date

File: packages/playground/src/utils/date.ts [hunk 16]
```diff
@@ -3,3 +3,27 @@ import moment from "moment";
...
+ * @description
+ * This function calculates the difference in days between the provided date and the current date.
+ * It then checks if this difference is less than or equal to 30 days.
+ *
+ * @example
+ * const date = new Date('2022-01-01');
+ * console.log(thirtyDaysInMilliseconds(date)); // Output: true or false depending on the current date
+ */
+
+export function thirtyDaysInMilliseconds(date: Date): boolean {
```
Comment (xmonader) [inline packages/playground/src/utils/date.ts]: the name is wrong for this one, this one returns true if the difference between two dates (one of them is fixed as the current one and the other one is the release date)

so should be

isDifferenceBetweenDates30Days(date1, date2) and when you use it pass date1 as the current and date2 gets the value of the release date

File: packages/playground/src/utils/date.ts [hunk 17]
```diff
@@ -12,14 +12,8 @@ export default function toHumanDate(timeInSeconds: number): string {
  * @returns {boolean} True if the release date is within the last 30 days, false otherwise.
  */
 
-export function isReleasedOverMon(releaseDate: Date, currentDate?: Date): boolean {
+export function isReleasedOverMon(releaseDate: Date, currentDate: Date): boolean {
   const thirtyDaysInMilliseconds = 24 * 60 * 60 * 1000;
```
Comment (xmonader) [inline packages/playground/src/utils/date.ts]: MillisecondsInADay

**Review by xmonader — APPROVED** [review]
